### PR TITLE
fix(table/index): default partitioned index ON at every level (#329)

### LINKED ISSUE #329: fix(table/index): flip partitioned-index default ON for all levels (follow-up to #296)
## Context

PR #325 closed #296 by landing Part 2 of its solution (TLI head/tail mirror). Part 1 of #296's solution and two of its acceptance items remain unimplemented:

> 1. **Force partitioned index as the default** for all SSTs. [...] Reduces blast radius of one bad sub-index from "whole SST" to "one partition".

The default `index_block_partitioning_policy` is still `PinningPolicy::new([false, false, false, true])` (only L3+ partitioned). L0/L1/L2 SSTs therefore continue to ship with a single full-index TLI: the TLI mirror from #296 Part 2 covers a head-vs-tail bit flip in the TLI itself, but a bit flip in a (single-block) full index has no within-block redundancy.

## Solution

1. Change the `Config` default for `index_block_partitioning_policy` to `PinningPolicy::all(true)` so every level emits a partitioned index. The `partitioned_filter_block_policy` default may also need to flip for symmetry — to be decided based on filter-block blast-radius analysis.

2. Add the integration test from #296's acceptance matrix that the mirror-only PR (#325) deferred: zero out a single sub-index partition, confirm that partition's data is unreachable but OTHER partitions still read OK. This is the test that justifies flipping the default — it demonstrates the blast-radius win.

3. Benchmark: confirm partitioned-default point-read regression < 5% on `db_bench` readrandom. #296 sets the bar; the bench result is the gate for merging the default flip.

## Acceptance

- [ ] `Config::default()` produces `index_block_partitioning_policy = PinningPolicy::all(true)` (or equivalent — partitioned across every level)
- [ ] Existing test fixtures updated where they relied on full-index behaviour (most will not — the read path tolerates both)
- [ ] New integration test: write a partitioned-index SST, zero out one sub-index partition's bytes, confirm reads from OTHER partitions still succeed and reads from the corrupted partition fail with a precise `BlockVerifyError::DataCorrupted` (or equivalent)
- [ ] `db_bench readrandom` (point reads) shows < 5% regression vs the previous full-index-at-L0/L1/L2 default; document the actual delta in the PR body
- [ ] If filter-block partitioning default should also flip for symmetry, address in the same PR; otherwise document the asymmetry rationale in `Config` docs

## Estimate

1-2d: the default flip itself is one line, but the test fixture cascade can be chunky depending on how many tests assert full-index-specific behaviour. The bench gate is the longest item if performance investigation is needed.

## Out of scope

- Further work on the TLI mirror itself (done in #325)
- Per-partition recovery / lazy repair (separate concern; the test here just asserts the existing recovery semantics are partition-isolated)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Closes #329 (the Part 1 follow-up to #296 — PR #325 shipped Part 2 / TLI head+tail mirror; this is the within-index blast-radius half).

- Flips `Config::default().index_block_partitioning_policy` from `PinningPolicy::new([false, false, false, true])` to `PinningPolicy::all(true)`. Every level now emits a partitioned index, so a bit flip inside one sub-index block only takes out the keys covered by that partition, not the entire SST.
- Adds the integration test deferred from #296's acceptance matrix: write a partitioned-index SST, walk the TLI, zero one middle sub-index block, verify intact partitions still read end-to-end and the corrupted one surfaces as an `Err` (not a silent miss).
- Leaves `filter_block_partitioning_policy` at the pre-#329 shape on purpose. A corrupt filter block produces a false negative (read short-circuits, caller misses an existing key) — a different hazard class from index corruption (which fails loudly). Flipping that default needs its own filter blast-radius / false-negative analysis; symmetry with index is not sufficient justification on its own. Rationale recorded in the inline `Config::default()` docstring.

## Bench result

`cargo bench --bench index_block -- table_point_read_index_layout` on the same data shape (4096 keys, 256 B data blocks, restart-interval 1):

| Layout | Point-read time |
|--------|-----------------|
| full   | 6.6963 us / 6.7466 us / 6.8256 us |
| partitioned | 3.8549 us / 3.8655 us / 3.8767 us |

Partitioned is ~43% **faster**, not slower. The TLI is smaller (only handles to partitions, not to every data block), and the sub-partition loaded for the actual lookup is smaller, so the two-stage binary search beats the single-stage search over a large full index. Well within the <5% regression gate from #329.

## Test plan

- [x] `cargo nextest run --workspace` — 1426 / 1426 passing
- [x] cargo clippy all-features all-targets -D warnings — clean
- [x] `cargo test --doc` — 43 / 43 passing
- [x] New `tests/partitioned_index_blast_radius.rs` exercises the within-SST isolation property the default flip relies on
- [x] Bench captured above; default flip is an improvement, not a regression

## Summary by CodeRabbit

* **Tests**
  * Added integration test to validate corruption isolation in partitioned index blocks, ensuring failed reads are limited to affected partitions while adjacent partitions remain accessible.

* **Chores**
  * Refactored benchmark code to support performance comparison across different index block layouts.
  * Updated default index block partitioning configuration and expanded related documentation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/coordinode-lsm-tree/pull/340?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/partitioned_index_blast_radius.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+    // The read that routes through the corrupted sub-index partition
+    // must surface the corruption as an error, NOT silently return
+    // None / wrong data. Either a block-header XXH3 failure or a
+    // ChecksumMismatch is acceptable.
+    let v_victim = tree.get(&victim_last_key, lsm_tree::MAX_SEQNO);
+    assert!(
+        v_victim.is_err(),
+        "read against corrupted sub-index partition must return an Err (got {v_victim:?})"
+    );
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/partitioned_index_blast_radius.rs:162]: Fixed in eb54baf4 — replaced is_err() with matches! over Error::ChecksumMismatch | Error::InvalidHeader. Inline comment records why both variants are accepted (zeroed header XXH3 mismatch vs structural-check failure on all-zero header).
